### PR TITLE
Added test for FileDownloadTrait

### PR DESCRIPTION
**Pull request by AIeksandr**
- Created File Download test.
- Added make command for files creating
- Created file field for article in d7 build

### CODE REVIEW

File: src/D7/FileDownloadTrait.php [hunk 1]
```diff
@@ -187,7 +191,7 @@ protected function fileDownloadOpenZip() {
       throw new \Exception('Downloaded file information does not have content type data.');
     }
 
-    if ($this->fileDownloadDownloadedFileInfo['content_type'] != 'application/octet-stream') {
+    if ($this->fileDownloadDownloadedFileInfo['content_type'] == 'application/octet-stream') {
```
Comment (AIeksandr) [inline src/D7/FileDownloadTrait.php]: Why we should check a zip header?

File: src/D7/FileDownloadTrait.php [hunk 2]
```diff
@@ -318,4 +322,26 @@ protected function fileDownloadGetTempDir() {
...
+        $file_source = $file_info[1];
+        $file_name = $file_info[2];
+        $image = file_get_contents($file_source);
+        file_save_data($image, 'public://' . $file_name, FILE_EXISTS_RENAME);
+
+        $node->{$key} = $file_name;
+      }
+    }
+  }
+
```
Comment (AIeksandr) [inline src/D7/FileDownloadTrait.php]: It test was firstly by priority, but the next(FileTrait) has "given files:" defenition. Should I remove this code and relation for it(Changes in MakeFile) and replace it with "given files:"? Cause I didn't know about this defenition before I started  next test

File: Makefile [hunk 3]
```diff
@@ -109,6 +109,7 @@ site-install:
 	$(call exec,docker-compose exec cli composer --working-dir=$(APP)/$(BUILD) install --prefer-dist)
 	$(call exec,docker-compose exec cli drush -r $(APP)/$(WEBROOT) si standard -y --db-url=mysql://drupal:drupal@$(MYSQL_HOST)/drupal --account-name=admin --account-pass=admin install_configure_form.enable_update_status_module=NULL install_configure_form.enable_update_status_emails=NULL)
 	$(call exec,docker-compose exec cli composer --working-dir=$(APP)/$(BUILD) drupal-post-install)
+	$(call exec,docker-compose exec cli cp -Rf $(APP)/$(FIXTURES)/$(EXAMPLE_FILES)/. $(APP)/$(WEBROOT)/$(EXAMPLE_FILES))
```
Comment (alexdesignworks) [inline Makefile]: this is not required - `tests/fixtures` are used - see explanation below

File: Makefile [hunk 4]
```diff
@@ -127,6 +128,7 @@ APP ?= /app
 WEBROOT ?= web
 BUILD ?= build
 FIXTURES ?= tests/behat/fixtures
+EXAMPLE_FILES ?= example_files
```
Comment (alexdesignworks) [inline Makefile]: not required

File: tests/behat/features/file-download.d7.feature [hunk 5]
```diff
@@ -0,0 +1,36 @@
+@d7
+Feature: Check that FileDownloadTrait for D7 works
+
+  Background:
+    Given I am logged in as a user with the "administrator" role
+    And article content:
+      | title                | field_file                                                   |
+      | [TEST] document page | file;example_files/example_document.txt;example_document.txt |
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: please update these to use `fixture` files as below `example_document.pdf` etc.  If you need a `.txt` file - please add it to `tests/fixtures` directory.

`tests/fixtures` is automatically discovered by DrupalExtension

File: src/D7/FileDownloadTrait.php [hunk 6]
```diff
@@ -318,4 +322,26 @@ protected function fileDownloadGetTempDir() {
...
   }
 
+  /**
+   * Before nodeCreate set field value for file.
+   *
+   * Field should be in format 'file;__file_source__;__file_name_.
+   *
+   * @beforeNodeCreate
+   */
+  public function nodeCreateAlter(EntityScope $scope) {
```
Comment (alexdesignworks) [inline src/D7/FileDownloadTrait.php]: this is not needed - the files are attached like so

```
 And article content:
      | title                | field_file                                                   |
      | [TEST] document page | example_file.txt |

```

and `example_file.txt` is located in `tests/fixtures`

File: tests/behat/features/file-download.d7.feature [hunk 7]
```diff
@@ -0,0 +1,36 @@
...
+    And article content:
+      | title           | field_file                                             |
+      | [TEST] zip page | file;example_files/example_files.zip;example_files.zip |
+
+  @api @download
+  Scenario: Assert that a file can be downloaded
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: Please use a step definition in scenario name for clarity

`Scenario: Assert "Then I download file from :url"`

File: tests/behat/features/file-download.d7.feature [hunk 8]
```diff
@@ -0,0 +1,36 @@
...
+    And I download file from "/example_files/example_document.txt"
+
+  @api
+  Scenario: Assert that a file can be downloaded by existing link
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: as above

File: tests/behat/features/file-download.d7.feature [hunk 9]
```diff
@@ -0,0 +1,36 @@
+@d7
+Feature: Check that FileDownloadTrait for D7 works
+
+  Background:
+    Given I am logged in as a user with the "administrator" role
+    And article content:
+      | title                | field_file                                                   |
+      | [TEST] document page | file;example_files/example_document.txt;example_document.txt |
+    And article content:
+      | title           | field_file                                             |
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: you do not need to repeat this twice - `test zip page` can go as a second line:

```
And article content:
      | title                | field_file                                                   |
      | [TEST] document page | file;example_files/example_document.txt;example_document.txt |
      | [TEST] zip page | file;example_files/example_files.zip;example_files.zip |
```

File: tests/behat/features/file-download.d7.feature [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+  @api @download
+  Scenario: Assert that a file can be downloaded
+    And I download file from "/example_files/example_document.txt"
+
+  @api
+  Scenario: Assert that a file can be downloaded by existing link
+    When I visit article "[TEST] document page"
+    Then I see download "Download example_document.txt" link "present"
+    Then I download file from link "Download example_document.txt"
+    And downloaded file contains:
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: Please add an additional step `And downloaded file does not contain:` and a test for it

File: tests/behat/features/file-download.d7.feature [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+    """
+    Some Text
+    """
+
+  @api
+  Scenario: Assert that an archive file can be downloaded
```
Comment (alexdesignworks) [inline tests/behat/features/file-download.d7.feature]: update this - as above

File: src/D7/FileDownloadTrait.php (hunk 2 above)
Comment (alexdesignworks) [reply src/D7/FileDownloadTrait.php]: yes.

`given files` actually creates files objects, but you still need to attach those objects to fields - see example below.

File: src/D7/FileDownloadTrait.php (hunk 1 above)
Comment (alexdesignworks) [reply src/D7/FileDownloadTrait.php]: sorry, I do not understand the question.

This is a ZIP-specific step, and `application/octet-stream` is a correct header for ZIP, unless I'm wrong  - please provide a link if I am

File: src/D7/FileDownloadTrait.php [hunk 12]
```diff
@@ -49,7 +51,9 @@ public function fileDownloadFrom($url) {
 
     /** @var Behat\Mink\Driver\BrowserKitDriver $driver */
     $driver = $this->getSession()->getDriver();
-    $cookies = $driver->getClient()->getCookieJar()->allValues($driver->getCurrentUrl());
+    $cookies = $driver->getClient()
```
Comment (alexdesignworks) [inline src/D7/FileDownloadTrait.php]: please revert this and fix config in your PHP storm to not wrap method calls.

File: src/D7/FileDownloadTrait.php (hunk 1 above)
Comment (AIeksandr) [reply src/D7/FileDownloadTrait.php]: Simple .zip hasn't this header. It has "application/zip". Anyway, I can't create archive with this header.

Comment (alexdesignworks) [reply src/D7/FileDownloadTrait.php]: oh, ok, can you please change this to accommodate both headers and if neither of them is present - throw exception.

File: tests/behat/fixtures/example_files/example_document.txt [hunk 13]
```diff
@@ -0,0 +1,2 @@
+Test Document
```
Comment (alexdesignworks) [inline tests/behat/fixtures/example_files/example_document.txt]: please remove this

**Pull request merged**